ext/zend_test/test.c

```diff
@@ -364,6 +365,67 @@ static ZEND_FUNCTION(zend_test_crash)
 	php_printf("%s", invalid);
 }
 
+zend_mm_heap* zend_test_heap;
+zend_mm_heap* zend_orig_heap;
+volatile uint32_t lineno = 0;
+
+static bool has_opline(zend_execute_data *execute_data)
+{
+	return execute_data
+		&& execute_data->func
+		&& ZEND_USER_CODE(execute_data->func->type)
+		&& execute_data->opline
+	;
+}
+
+#pragma GCC push_options
+#pragma GCC optimize("O0")
+void * __attribute__((optnone)) zend_test_custom_malloc(size_t len)
+{
+	if (has_opline(EG(current_execute_data))) {
+		lineno = EG(current_execute_data)->opline->lineno;
+	}
+	return _zend_mm_alloc(zend_orig_heap, len);
+}
+
+void __attribute__((optnone)) zend_test_custom_free(void *ptr)
+{
+	if (has_opline(EG(current_execute_data))) {
+		lineno = EG(current_execute_data)->opline->lineno;
+	}
+	_zend_mm_free(zend_orig_heap, ptr);
+}
+
+void * __attribute__((optnone)) zend_test_custom_realloc(void * ptr, size_t len)
+{
+	if (has_opline(EG(current_execute_data))) {
+		lineno = EG(current_execute_data)->opline->lineno;
+	}
+	return _zend_mm_realloc(zend_orig_heap, ptr, len);
+}
+#pragma GCC pop_options
+
+static ZEND_FUNCTION(zend_test_observe_opline_in_zendmm)
+{
```

Comment:
Would be great if this wasn't a magic 4096 value.

Reply:
I used this magic number, as I do not know of a way to get the actuall `zend_mm_heap` size as this is a private struct defined in `Zend/zend_alloc.c`. The only thing publicly available is the pointer to that struct which is defined in `Zend/zend_alloc.h`. If you have a way to find out the actual size, I will update, otherwise we could store that magic number in a const?

Reply:
I see. As this is test code I don't care that much, just put a comment to clarify I guess.

Reply:
done